pkg/quota/evaluator/core/persistent_volume_claims.go

```diff
@@ -51,68 +54,127 @@ func listPersistentVolumeClaimsByNamespaceFuncUsingClient(kubeClient clientset.I
+	constraintsFunc := makePersistentVolumeClaimConstraintsFunc(usageFunc)
 	listFuncByNamespace := listPersistentVolumeClaimsByNamespaceFuncUsingClient(kubeClient)
 	if f != nil {
 		listFuncByNamespace = generic.ListResourceUsingInformerFunc(f, schema.GroupResource{Resource: "persistentvolumeclaims"})
 	}
-
 	return &generic.GenericEvaluator{
 		Name:              "Evaluator.PersistentVolumeClaim",
 		InternalGroupKind: api.Kind("PersistentVolumeClaim"),
-		InternalOperationResources: map[admission.Operation][]api.ResourceName{
-			admission.Create: allResources,
+		Operations:        []admission.Operation{admission.Create},
+		MatchedResourceNamesFunc: func(items []api.ResourceName) (result []api.ResourceName) {
+			for _, item := range items {
+				if quota.Contains(fixedResources, item) {
+					result = append(result, item)
+					continue
+				}
+				// match pvc resources scoped by storage class (<storage-class-name>.storage-class.kubernetes.io/<resource>)
+				for _, resource := range fixedResources {
+					byStorageClass := storageClassSuffix + string(resource)
+					if strings.HasSuffix(string(item), byStorageClass) {
+						result = append(result, item)
+						break
+					}
+				}
+			}
+			return result
 		},
-		MatchedResourceNames: allResources,
-		MatchesScopeFunc:     generic.MatchesNoScopeFunc,
-		ConstraintsFunc:      PersistentVolumeClaimConstraintsFunc,
-		UsageFunc:            PersistentVolumeClaimUsageFunc,
-		ListFuncByNamespace:  listFuncByNamespace,
+		MatchesScopeFunc:    generic.MatchesNoScopeFunc,
+		ConstraintsFunc:     constraintsFunc,
+		UsageFunc:           usageFunc,
+		ListFuncByNamespace: listFuncByNamespace,
 	}
 }
 
-// PersistentVolumeClaimUsageFunc knows how to measure usage associated with persistent volume claims
-func PersistentVolumeClaimUsageFunc(object runtime.Object) api.ResourceList {
-	result := api.ResourceList{}
-	var found bool
-	var request resource.Quantity
-
-	switch t := object.(type) {
+func toInternalPersistentVolumeClaimOrDie(obj runtime.Object) *api.PersistentVolumeClaim {
+	pvc := &api.PersistentVolumeClaim{}
+	switch t := obj.(type) {
 	case *v1.PersistentVolumeClaim:
-		request, found = t.Spec.Resources.Requests[v1.ResourceStorage]
+		if err := v1.Convert_v1_PersistentVolumeClaim_To_api_PersistentVolumeClaim(t, pvc, nil); err != nil {
+			panic(err)
+		}
 	case *api.PersistentVolumeClaim:
-		request, found = t.Spec.Resources.Requests[api.ResourceStorage]
+		pvc = t
 	default:
-		panic(fmt.Sprintf("expect *api.PersistenVolumeClaim or *v1.PersistentVolumeClaim, got %v", t))
+		panic(fmt.Sprintf("expect *api.PersistentVolumeClaim or *v1.PersistentVolumeClaim, got %v", t))
 	}
+	return pvc
+}
 
+// PersistentVolumeClaimUsageFunc knows how to measure usage associated with persistent volume claims
+func PersistentVolumeClaimUsageFunc(object runtime.Object) api.ResourceList {
+	result := api.ResourceList{}
+	pvc := toInternalPersistentVolumeClaimOrDie(object)
```

Comment:
It's not reasonable to issue a panic (which kills the apiserver by default) on a quota evaluation error.  I understand why you're doing a cast/conversion, but I think that means this should return an error.

I could also accept a link to where we always ensure this will cast or convert successfully, but its probably structurally better to allow an error.